Add security group subcollection actions

tzumainn opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1487103

jntullo commented on app/controllers/api/subcollections/security_groups.rb:
```diff
@@ -4,6 +4,30 @@ module SecurityGroups
+
+        begin
+          message = "Adding security group #{security_group} to #{parent.name}"
+          task_id = queue_object_action(parent, message, :method_name => "add_security_group", :args => [security_group])
```
I see the `query_resource` checks to see if it responds to `:security_groups` - I am assuming this means not all Instances have security groups, so should it be checking to see if the parent also responds to `add_security_group` before the action is queued? Same for the `remove_resource`

jntullo commented on spec/requests/instances_spec.rb:
```diff
@@ -616,4 +616,95 @@ def update_raw_power_state(state, *instances)
+        'resources' => [
+          { 'href' => api_instance_security_group_url(nil, instance, @security_group) }
+        ]
+
```
nit, but is this just an empty line?

tzumainn commented:
@jntullo thanks for the review! I've made the requested changes, let me know if you think there should be more!

jntullo commented on app/controllers/api/subcollections/security_groups.rb:
```diff
@@ -4,6 +4,32 @@ module SecurityGroups
+
+      def security_groups_add_resource(parent, _type, _id, data)
+        security_group = data["security_group"]
+        raise "Cannot add #{security_group} to #{parent.name}" unless parent.respond_to? :add_security_group
```
Any reason why this raise is outside of the rescue block? Returning action results for everything plays nicer with bulk actions.

abellotti replied:
was gonna ask, we'd need specs for adding/removing in bulk.

jntullo commented:
@tzumainn can you add tests for the new check you did for whether it responds to adding or removing security groups?

tzumainn commented:
@jntullo thanks for the additional review! I've moved the raises statements and added the tests. Let me know what you think!

@abellotti this operation can't be done in bulk - let me know if I'm misunderstanding your comment?

abellotti commented:
why can't it be done in bulk ?  instances can have multiple security groups unless I am missing something. Standard signature would be as follows:

```
POST /api/instances/:id/security_groups

{
  "action" : "add",
  "resources" : [
      { "security_group" : "security_group_A" },
      { "security_group" : "security_group_B" }
  ]
}
```

same with the remove signature,

```
POST /api/instances/:id/security_groups

{
  "action" : "remove",
  "resources" : [
      { "security_group" : "security_group_D" },
      { "security_group" : "security_group_E" }
  ]
}
```

tzumainn commented:
@abellotti Ah, thanks for the explanation; I've added bulk tests. Thanks!

tzumainn commented:
the remaining style issues look like false positives due to nested hashes

tzumainn commented:
@abellotti let me know if you'd like further changes - thanks!

abellotti commented:
the one change that I'd like is to use "name" instead of "security_group" in the code/specs.

```
{
   "action" : "add",
   "resources" : [
      { "name" : "security_group_name1" },
      { "name" : "security_group_name2" }
   ]
}
```

would allow us to support href's, id's, later.

also, if you could take care of rubocop warnings (I saw extra spaces before =>, and params not aligned with first method argument for example).

tzumainn commented:
@abellotti fixed, thanks for the review! I think the last rubocop issues are incorrect - things look correctly aligned to me - but let me know if I'm wrong!

abellotti commented:
Thanks @tzumainn for the API Enhancement 👍
